### PR TITLE
Generate dockerfiles dynamically

### PR DESCRIPTION
**Pull request by superm1**
Rather than needing to remember to go update 4 dockerfiles, update one matrix and all the dockerfiles are updated accordingly.

### CODE REVIEW

File: contrib/ci/generate_dockerfile.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/python3
```
Comment (hughsie) [inline contrib/ci/generate_dockerfile.py]: maybe some copyright and license here?

File: contrib/ci/generate_dockerfile.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import os
+import sys
+
+directory = os.path.dirname(sys.argv[0])
```
Comment (hughsie) [inline contrib/ci/generate_dockerfile.py]: maybe checking len(argv)?

**Comment by hughsie** [comment]
I don't know if this proposal is crazy or genius. I guess the difference between the two are success, and all the builds seem to be passing... :)

File: contrib/ci/generate_dockerfile.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+TARGET=os.getenv('OS')
+
+if TARGET == '':
+    print("Missing OS environment variable")
+    sys.exit(1)
+
+with open (os.path.join(directory,"dependency_matrix.csv"), 'r') as rfd:
```
Comment (hughsie) [inline contrib/ci/generate_dockerfile.py]: Not 100% about the name -- dependencies.txt or depends.txt might be better.

Comment (superm1) [reply contrib/ci/generate_dockerfile.py]: OK.

**Comment by superm1** [comment]
If I could manage to generate contrib/debian/control and contrib/fwupd.spec dynamically I'd go with genius.  There's some extra metadata needed there to make that happen, but I want all this in one place.

File: contrib/ci/generate_dockerfile.py (hunk 2 above)
Comment (superm1) [reply contrib/ci/generate_dockerfile.py]: Even when you're always grabbing 0? to get the path?  That should always be there.

File: contrib/ci/generate_dockerfile.py (hunk 1 above)
Comment (superm1) [reply contrib/ci/generate_dockerfile.py]: OK, will add some.

File: contrib/ci/generate_dockerfile.py (hunk 2 above)
Comment (hughsie) [reply contrib/ci/generate_dockerfile.py]: Yup, you're right, sorry.

**Comment by hughsie** [comment]
Also, do you want to squash this down to 1 commit pls.

**Comment by superm1** [comment]
Yeah I'll squash it to one commit when I merge.  Once I make those fixes on board with this approach though?

**Comment by hughsie** [comment]
Initially I was "err, so we build the source files for a meta-generator of a virtual build system" and maybe wondered if we were going too deep down the rabbithole and tying ourselves to github+travis. That said, maintaining the current matrix is tricky indeed, and so this seems like a pragmatic solution. It does seem a shame we can't use the output of the toplevel meson file as an input to the generator, but that's only going to be a subset of what we need for each builder. From a 30,000ft point of view we want to say install the thing that installs `valgrind.h`, `libsoup.pc` etc and I wonder if providing a mapping for each distro might be better than just duplicating the requirements in two places.

**Comment by hughsie** [comment]
I was thinking about this a lot last night. I do think generating the Dockerfile-foo files is a good idea, however I'm not really sold on the magic python script that conditonalizes a lot of the Dockerfile system-specific bits -- it's just putting distro-specific hardcoding in the python file rather than in the Dockerfile directly -- but the python script that creates the list of package deps *does* make a lot of sense to me.

Perhaps we could have contrib/ci/Dockerfile-fedora.in which is preprocessed using your python script to contrib/ci/Dockerfile-fedora and inside it we just have `%%%REPLACE_ME_WITH_RUN_COMMANDS%%%` which gets replaced with the package manger command (e.g. `apt-get`) and also the list of packages from your mega-matrix. The python script would just take the input file, the output name and the distro id and then just replace the `%%%` bit and we could just use `custom_target` in the meson build file. The only fly in the ointment would be if the Dockerfile is needed before meson is run.  :/ I'm no expert here.

I'd be much happier with less magic, and I think allows us to easily drop in a new distro with adding a new Dockerfile-foo file with the boilerplate and then just a new column in your matrix.

What do you think?

**Comment by superm1** [comment]
Assuming CI works on that, take a look at what I did instead.  I think that's closer along the lines of what you have in mind and is achievable.

`meson `doesn't exist outside of the docker container, so I don't think any fancy stuff can be done with a custom target in that regard.

**Comment by hughsie** [comment]
That's exactly what I had in mind. Squash, and then LGTM. Thanks!

**Pull request merged**